Update language documentation (#232)
candy-lang/candy: language.md

- **Candy is expressive.**
  You can be flexible with how you model your data in Candy.
  We aim to have a reasonable concise syntax to express common patterns.

Candy aims to blur the line between compile-time and run-time, for example, by replacing compile-time types with edit-time fuzzing.

This document assumes you're familiar with other programming languages and covers most of what's planned for Candy.
Note that not all of the features described here are implemented or even finalized.

- [Basic Syntax](#basic-syntax)
- [Values](#values)
  - [Integers](#integers)
  - [Texts](#texts)
  - [Symbols](#symbols)
  - [Structs](#structs)
  - [Closures](#closures)
  - [More?](#more)
  - [Channel Ends](#channel-ends)
- [Variables](#variables)
- [Functions](#functions)
- [Modules](#modules)
- [Comments](#comments)
- [Panics](#panics)
- [Needs](#needs)
- [Pattern matching](#pattern-matching)
- [Concurrency](#concurrency)
  - [Channels](#channels)
- [Packages](#packages)
- [Environment](#environment)
- [Interoperability with other languages](#interoperability-with-other-languages)
  - [Add to the Environment](#add-to-the-environment)
  - [Contain Pure Code](#contain-pure-code)

> - [Deploying code](#deploying-code)

Reviewer: Some headers use title case, others don't

## Basic Syntax

Candy's syntax is inspired by Elm and Haskell.

Source code is stored in plain text files with a `.candy` file extension.

> Line [comments](#comments) start with `#` and end at the end of the line:

Reviewer: As we only have line comments and it's explained that they go to the end of the line, I don't think we have to write “Line comments”

```suggestion
[Comments](#comments) start with `#` and end at the end of the line:
```

```candy
# This is a comment.
```

Naming rules are similar to other programming languages.

> Identifiers start with a lowecase letter and may contain letters or digits.

Reviewer: ```suggestion
Identifiers start with a lowercase letter and may contain letters or digits.
```

Case is sensitive.

```candy
hi
camelCase
abc123
```

Newlines and indentation are both meaningful in Candy.
Newlines separate code expressions.
Indentation always consists of two spaces and is used to group several expressions in [scopes](#variables).
Scopes evaluate to their last expression.

```candy
foo = 42
bar =
  baz = 3
  5
# bar is now 5
```

## Values

All data in your program is composed of values.
Values can be created through literals, expressions that evaluate to a value.

> All values are immutable – once created, they do not change.

Reviewer: ```suggestion
All values are immutable – once created, they do not change.
```

The number `3` is always the number `3`.
The string `"frozen"` can never have its character array modified in place.

There only exist few types of values in Candy.

### Integers

Integers are arbitrarily large, whole numbers.
You write them in decimal.

```candy
0
42
123456789012345678901234567890
```

TODO: Different radixes

### Texts

Texts are Unicode strings.
They start and end with double quotes (`"`) and can span multiple lines if they're indented.

You can also start texts with any number of single quotes (`'`) followed by a double quote (`"`).
This so-called meta-text can only be ended with a double quote and the same number of single quotes that it started with.

> Finally, you can use curly braces (`{}`) containing a value to insert a stringified version of the value into the string at that position.

Reviewer: I am thinking about only allowing strings to be inserted, meaning you have to convert your value to a string before inserting it in an interpolation. The reason is that the debug-formatted string is not always what you want and it can be easy to forget to call your custom stringification method if inserting the raw value also works. We could also think about one special character in interpolations as a shortcut for the debug string, though I don't want to add format strings with more features.

```candy
"Hello!"
"A somewhat
  long
  text."
'"This is a meta text, where you can use " inside the text."'
''"This is a double-meta text, allowing you to use "' inside it without ending it."''
"Some {interpolation}."

> '"In meta texts, {{interpolation}} requires more curly braces, otherwise the values are {not interpolated}."'

Reviewer: ```suggestion
'"In meta texts, {{interpolation}} requires more curly braces; otherwise, the values are {not interpolated}."'
```

### Closures

Closures are pieces of code that can be executed.

```candy
identityFunction = { argument -> argument }
longClosure = { foo ->
  ...
}
```

### More?

TODO: Lists? Sets?

- lists: [1, 2, 3]
- sets: Clojure has %{ value }
  - or like Toit? {hey, you, there} for set, empty map is {:}

### Channel Ends

Channels ends allow you to interact with a [channel](#concurrency).
There are receive ends and send ends to receive and send data from a channel, respectively.

## Variables

Variables are named slots for storing values.
You define a new variable using the equals sign `=`, like so:

```candy
foo = "Hello!"
```

This creates a new variable `foo` in the current scope and initializes it with the result of the expression following the `=`.
Once a variable has been defined, it can never be reassigned.
You can access variables by their name.

```candy
foo = "Hi!"
bar = foo  # bar = "Hi!"
```

Variables only exist until the end of the scope they're defined in.

```candy
foo =
  bar = hello  # error because hello doesn't exist yet
  hello = 5
  4
bar = hello  # error because hello doesn't exist anymore
```

> Variables at the top level of a file are visible to the module system.

Reviewer: ```suggestion
Variables at the file's top level are visible to the module system (“public”) if they are declared using `:=`.
```

All other variables are local.

> Declaring a variable with the same name as another simply shadows that variable:

Reviewer: ```suggestion
Declaring a variable with the same name as another simply shadows that variable, though that's not allowed for public variables.
```

From that point forward, the name will refer to the new variable.
Note that this is different from reassigning to a variable.

```candy
foo = 5
foo = 3
# there is no way to get to the 5
bar =
  foo = 4
# here, foo is still 3
```

## Functions

> Functions can either be defined using closure literals or by writing them as parameterized variables with arguments in front of the `=`.

Reviewer: ```suggestion
Functions can either be defined using closure literals or by writing them as parameterized variables with arguments in front of the `=` or `:=`.
```

Both representations are equivalent with respect to what they do during runtime.

```candy

> # Both these definitions are equivalent.

Reviewer: Already stated above about their behavior during runtime. During compile-time, however, they differ with regards to fuzzing.

For example, a Candy project might look like this:

```
main.candy
green/
  .candy
  brown.candy
red/
  .candy
  yellow/
    .candy
    purple.candy
  blue.candy
```

This directory structure corresponds to the following module hierarchy:

```
main        # from main.candy
green       # from green/.candy
  brown     # from green/brown.candy
red         # from red/.candy
  yellow    # from red/yellow/.candy
    purple  # from red/yellow/purple.candy
  blue      # from red/blue.candy
```

Inside a module, top-level variable definitions can use a `:=` instead of `=` to export a variable.
In each module, there automatically exists a `use` function that will import other modules from the module tree.
You pass it a text that describes what module to import.

```
# inside red/yellow/.candy

foo = use ".purple"  # imports the purple child module
foo = use "..blue"   # imports the blue sibling module
foo = use "...green" # imports the green parent module
```

Each additional dot at the beginning symbolizes a navigation one level up.
The possible multiple dots are followed by the name of the module to import.

> Note that you can't navigate further than one level in – for example, the `yellow` module can't import the `brown` module, only its parent module `green`.

Reviewer: ```suggestion
Note that you can't navigate further than one level in – for example, the `yellow` module can't import the `brown` module, only its parent module `green`.
```

The `use` call evaluates the given module and returns a struct containing all its exported definitions (variables and functions using `:=`).

```candy
# inside green/brown.candy

foo = 3
bar := foo
baz a := a
```

```candy
# inside green/.candy

brown = use ".brown"

# equivalent:
brown =
  foo = 3
  bar = foo
  baz a = a
  [
    Bar: foo,
    Baz: bar,
  ]

> # equivalent:

Reviewer: Except for the fuzzing behavior

brown = [
  Bar: 5,
  Baz: { a -> a },
]
```

> The `useAsset` also allows you to import arbitrary non-Candy-files part from your module hierarchy.

Reviewer: ```suggestion
The `useAsset` also allows you to import arbitrary non-Candy files that are part of your module hierarchy.
```

In some cases, it makes more sense to express some data in other formats.
For example, you might want to store user-facing translations for your program in a JSON file.

```
main.candy
translations.json
```

```
# inside main.candy

> translations = json.parse(useAsset "..translations.json")

Reviewer: ```suggestion
translations = json.parse (useAsset "..translations.json")
```

translations.helloWorld
```

Changes to these files are also tracked by the Candy tooling and autocompletions and hints will update accordingly.

## Comments

TODO: Write something including doc comments

## Panics

Candy programs can panic, causing them to crash.
Contrary to crashes in other programming languages, it's always programmatically clear which part of the code is at fault.

> For example, in Rust, if a function panics you have to look at its documentation to understand if the panic is your fault or not:

Reviewer: ```suggestion
For example, in Rust, if a function panics, you have to look at its documentation to understand if the panic is your fault or not:
```

```candy
foo a =
  needs (core.int.is a) "life's not fair"

foo Hey  # Calling `foo Hey` panics because life's not fair.
```

The editor tooling will analyze your functions and try them out with different values.
If an input crashes in a way that your code is at fault, you will see a hint.

```candy
mySqrt a =               # If you pass `a = -1`,
  needs (core.int.is a)  # this needs succeeds because `core.int.is -1 = True`,
  core.int.sqrt a        # but calling `core.int.sqrt -1` panics because sqrt only works on non-negative integers. If you think this should be different, check out the `ComplexNumbers` package.
```

## Pattern matching

TODO: Write something

TODO: Irrefutable and refutable patterns

## Concurrency

Candy supports a lightweight version of threads called *fibers*.
To enforce structured concurrency, they can only be spawned using a special concurrency object called the *nursery*.
In the following code, the surrounding call to `core.parallel` only exits when all fibers inside have completed.

```candy
core.parallel { nursery ->
  core.fiber.spawn nursery { print "Banana" }
  core.fiber.spawn nursery { print "Kiwi" }
  # Banana and Kiwi may print in any order
}

> print "Peach"  # Always prints after the others

Reviewer: I don't think this simple `print` call would work with capabilities

```candy
channel = core.channel.new 5  # creates a new channel with capacity 5
sender = channel.sendEnd
receiver = channel.receiveEnd

core.channel.send sender Foo
core.channel.send sender Bar
core.channel.receive receiver  # Foo
core.channel.receive receiver  # Bar
```

There is no guaranteed ordering between messages sent by multiple fibers, but messages coming from the same fiber are guaranteed to arrive in the same order they were sent.

All channel operations are blocking.
Thus, channels also function as a synchronization primitive and can generate *backpressure*.

```candy
core.parallel { nursery ->
  channel = core.channel.new 3
  core.fiber.spawn nursery {
    loop {
      core.channel.send channel.sendEnd "Hi!"
    }
  }
  core.fiber.spawn nursery {
    loop {
      print (core.channel.receive channel.receiveEnd)
    }
  }
}
```

In this example, if the printing takes longer than the generating of new texts to print, the generator will wait for the printing to happen.
At most 3 texts will exist before being printed.

TODO: Write about async await as soon as they are in the Core package

## Packages

TODO: Write something

## Environment

> At some point, your Candy program needs to have side-effects – otherwise, it's just heating up your CPU.

Reviewer: ```suggestion
At some point, your Candy program needs to have side effects – otherwise, it's just heating up your CPU.
```

If you develop for a new platform or want to enable more functionality in the native platform, we will have some way of plugging a new part of native code into the runtime that can make its own capabilities available on the environment passed to `main`.

For example, on a microcontroller, the stdout capability doesn't make sense.
Instead, you might wave a pin capability that allows you to modify the voltage of the hardware pins.

### Contain Pure Code

If you want to use existing code that implements pure functions, it can make sense to compile the existing code into WebAssembly (WASM).
You can put the resulting WASM module in your Candy module hierarchy, call `useAsset` with that file, and pass the data to a WASM runtime that we'll write in Candy when we get to it.

This approach moves the native code entirely into the Candy domain, so the Candy compiler can also reason about the WASM code.
For example, if you call a function of your WASM module only with inputs known at compile-time, the Candy compiler may execute those calls directly and not include the original WASM code in the binary at all.

## Deploying code

This chapter is especially experimental and spitball-y.

> Similar to how Zig build scripts work, we may support having a `build.candy` file that contains information about how to compile and optimize your code.

Reviewer: What would be written in there? Should this really be something customizable?

Next to the interpreted VM, we plan to compile to LLVM or WASM.

One idea we had is to let you provide a custom scoring function in the build script instead of having binary options like "optimize for speed" or "optimize for performance".
This scoring function could get used by the optimizer to choose which paths to take.

> For example, you could formulate a build where you're okay with a resulting binary blowup of 1 KiB per 10 ms of saved time in some annotated performance-crititcal section.

Reviewer: ```suggestion
For example, you could formulate a build where you're okay with a resulting binary blowup of 1  KiB per 10 ms of saved time in some annotated performance-critical section.
```

> Or when developing for an embedded device with limited storage capacity, you might want to generate a binary that fits in the limit but is otherwise as fast as possible.

Reviewer: ```suggestion
Or, when developing for an embedded device with limited storage capacity, you might want to generate a binary that fits within the limit but is otherwise as fast as possible.
```

[
  Name: "Candy",
  Foo: 42,
  "TextKey": 4,
  3: 2,
]
```

> To use numbers as keys, you can also omit the keys of 
> 

Reviewer: ```suggestion
```

> Omitting keys is equivalent to using zero-indexed numbers as keys:
> 
> ```candy
> [Hello, World]
> 
> # equivalent:
> [0: Hello, 1: World]

Reviewer: ```suggestion
Omitting keys is equivalent to using zero-indexed numbers as keys. You can also add fields with explicit keys after these:

```candy
[Hello, World, Foo: Bar]

# equivalent:
[0: Hello, 1: World, Foo: Bar]
```

}
```

### Channel Ends

TODO: Or are they called ports?

Channels ends allow you to interact with a [channel](#concurrency).
There are receive ends and send ends to receive and send data from a channel, respectively.

### More?

TODO: Lists? Sets?

> - lists: `[1, 2, 3]`

Reviewer: Were just merged :)

## Pattern Matching

Candy supports structural pattern matching using the match operator `%`.

```candy

> match (foo 5) %

Reviewer: This looks like `maybe` is also part of the necessary syntax. Maybe call a different function?

  [Ok, value] -> ...
  [Error, errorValue], core.int.isEven errorValue -> ...
  _ -> ...
```

Here, each indented line after the match operator represents a match case.
Each case can match based on the pattern as well an optional condition separated by a comma.
The first matching case is executed.

> If no case matches, you code panics.

Reviewer: ```suggestion
If no case matches, your code panics.
```

If you're sure about the structure of a value, you can also use patterns on the left-hand side of an assignment.
These are called irrefutable patterns.
Again, if the pattern doesn't match, the code panics.

```candy

> [a, b] = myStruct

Reviewer: ```suggestion
[a, b] = myList
```
